aspnetcore/blazor/javascript-interoperability/static-server-rendering.md

```diff
@@ -0,0 +1,186 @@
+
+`wwwroot/BlazorPageScript.lib.module.js`:
+
+```javascript
+const pageScriptInfoBySrc = new Map();
+
+function registerPageScriptElement(src) {
+  if (!src) {
+    throw new Error('Must provide a non-empty value for the "src" attribute.');
+  }
+
+  let pageScriptInfo = pageScriptInfoBySrc.get(src);
+
+  if (pageScriptInfo) {
+    pageScriptInfo.referenceCount++;
+  } else {
+    pageScriptInfo = { referenceCount: 1, module: null };
+    pageScriptInfoBySrc.set(src, pageScriptInfo);
+    initializePageScriptModule(src, pageScriptInfo);
+  }
+}
+
+function unregisterPageScriptElement(src) {
+    if (!src) {
+        return;
+    }
+
+    const pageScriptInfo = pageScriptInfoBySrc.get(src);
+    if (!pageScriptInfo) {
+        return;
+    }
+
+    pageScriptInfo.referenceCount--;
+}
+
+async function initializePageScriptModule(src, pageScriptInfo) {
+  if (src.startsWith("./")) {
+    src = new URL(src.substr(2), document.baseURI).toString();
+  }
+
+  const module = await import(src);
+
+  if (pageScriptInfo.referenceCount <= 0) {
+    return;
+  }
+
+  pageScriptInfo.module = module;
+  module.onLoad?.();
+  module.onUpdate?.();
+}
+
+function onEnhancedLoad() {
+  for (const [src, { module, referenceCount }] of pageScriptInfoBySrc) {
+    if (referenceCount <= 0) {
+      module?.onDispose?.();
+      pageScriptInfoBySrc.delete(src);
+    }
+  }
+
+  for (const { module } of pageScriptInfoBySrc.values()) {
+    module?.onUpdate?.();
+  }
+}
+
+export function afterWebStarted(blazor) {
+  customElements.define('page-script', class extends HTMLElement {
+    static observedAttributes = ['src'];
+
+    attributeChangedCallback(name, oldValue, newValue) {
+      if (name !== 'src') {
+        return;
+      }
+
+      this.src = newValue;
+      unregisterPageScriptElement(oldValue);
+      registerPageScriptElement(newValue);
+    }
+
+    disconnectedCallback() {
+      unregisterPageScriptElement(this.src);
+    }
+  });
+
+  blazor.addEventListener('enhancedload', onEnhancedLoad);
+}
+```
+
+For the preceding module:
+
+* In the `initializePageScriptModule` function:
+  * A relative path is normalized by by making it an absolute URL with document's app base path.
+  * If `pageScriptInfo.referenceCount <= 0`, all `<page-script>` elements with the same `src` were unregistered while the module was loading.
+* In the `onEnhancedLoad` function:
+  * Start by invoking `onDispose` on any modules that are no longer referenced.
+  * Invoke `onUpdate` on the remaining modules.
+* In the `afterWebStarted` function, `attributeChangedCallback` is used instead of `connectedCallback` because a page script element might be reused between enhanced navigation operations. For more information, see [Using custom elements: Custom element lifecycle callbacks (MDN documentation)](https://developer.mozilla.org/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks).
```

Comment:
Are there other examples in the docs where we go in this much detail about a code sample? I think this makes sense when the sample demonstrates a framework feature, but I'd imagine this is just code people are mostly going to copy and paste. For example, [this sample](https://learn.microsoft.com/aspnet/core/blazor/fundamentals/signalr?view=aspnetcore-7.0#automatically-refresh-the-page-when-server-side-reconnection-fails) is somewhat lengthy but we don't comment much about what it's doing.

Reply:
Only placed because you had them in the code sample, where they don't get localized. I'll strip this bit out now.